### ISSUE TITLE
Validatation failures, next round

### ISSUE THREAD

**Issue by DavidBrainard** [isetbio/isetbio#67]
Just ran a full validation.  Two failures:

1) v_DisplayColorConversion.  This seems pretty minor.  It passes the FAST validation, but fails the FULL.  The failure is a very small difference in the value of variable ptbLMSIsomerizations between the ground truth (written yesterday by Nicolas) and what my machine produces.  The difference is on the order of 10^-12, for values that are on the order of 10^3.  I'm guessing these are real precision differences between what Nicolas's machine (ground truth FAST and FULL both written by Nicolas yesterday and are in sync with each other) and mine do, and that the FAST passes because we round to lower precision before hashing.  If this is indeed the case, we should probably increase the tolerance for the FULL comparison.

2)  v_HumanRetinalIlluminance580nm.  Here both FAST and FULL fail.  The FAST validation was written by Brian on July 16, the FULL by me on 16 April.  The FULL failure lists three problems.  a) The optics.OTF.OTF used to be a 2-D matrix but is now a 3-D matrix.  The oi.optics.OTF.wave used to be a [1 by 1] matrix but is now a [31 by 1] matrix.  And scene.illuminant.spectrum.wave no longer matches what it used to have.

I'm thinking these are related to real changes made in the code since 16 April and since 16 July, since Brian recently posted about fussing with the wavelength sampling stuff in oi at least.  And it may be that the result of this is that we used to run with a single wavelength in this validation script but are now running with a polychromatic OTF.  But I'm not sure enough to just blotto the old ground truth data and regenerate new data.

The output of the validation run for these two functions is below.

Also, these scripts print a message that there are some figures that show the differences, but no such figures have appeared.  I'm guessing no figures appeared because I don't have some flag set, but if so then perhaps the code that produces the message about the availability of the figures should only do so when the figures are being generated.

I will try the validate and publish stuff once we straighten these out.

*******************************

[  5] 'v_DisplayColorConversion'
	Internal validation  : PASSED
	Run-time status      : no exception raised
	Full validation file : contains 1 instance of historical data.
	Ground truth  file   : /Users/Shared/Matlab/Toolboxes/isetbio/validation/data/fast/display/v_DisplayColorConversion_FastGroundTruthDataHistory.mat
	Fast validation      : PASSED against ground truth data of 30-Jul-2015 01:59:35.
	 > Ground truth info :                   Ithaka.local / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'nicolas'
	 > Local host info   :            ray.PSYCH.UPENN.EDU / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'dhb'
	Data hash key        : c1402bc8030ff0651bd18b40ed7729c90f7d31b96a892e6f878e7b9f5eb33bd5
	---------------------------------------------------------------------------------------------------------------------------------
	Full validation file : contains 1 instance of historical data.
	Ground truth  file   : /Users1/Shared/Dropbox/ISETBIOFullValidationData/display/v_DisplayColorConversion_FullGroundTruthDataHistory.mat
	Full validation      : FAILED against ground truth data of 30-Jul-2015 01:59:37.
	 > Ground truth info : Ithaka.local                   / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'nicolas'
	 > Local host info   : ray.PSYCH.UPENN.EDU            / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'dhb'
	[data mismatch  1]   : Max difference between 'groundTruthData.conversionData.ptbLMSIsomerizations' and 'validationData.conversionData.ptbLMSIsomerizations' (1.81899e-12) is greater than the set tolerance (1.11022e-13). See figure named: ''groundTruthData.conversionData.ptbLMSIsomerizations' vs. 'validationData.conversionData.ptbLMSIsomerizations''
 	Note (*)             : script does not store any extra data.
	Validation report    : 
		Compare isetbio and PTB display color conversion.
		Red inversion comparison is within the specified tolerance (0.001) of zero.
		Green inversion comparison is within the specified tolerance (0.001) of zero.
		Blue inversion comparison is within the specified tolerance (0.001) of zero.
		PTB red settings->primary->settings check is within the specified tolerance (0.001) of zero.
		PTB green settings->primary->settings check is within the specified tolerance (0.001) of zero.
		PTB blue settings->primary->settings check is within the specified tolerance (0.001) of zero.
		L isomerization comparison is within the specified tolerance (0.01) of zero.
		M isomerization comparison is within the specified tolerance (0.01) of zero.
		S isomerization comparison is within the specified tolerance (0.01) of zero.

[  8] 'v_HumanRetinalIlluminance580nm'
	Internal validation  : PASSED
	Run-time status      : no exception raised
	Full validation file : contains 1 instance of historical data.
	Ground truth  file   : /Users/Shared/Matlab/Toolboxes/isetbio/validation/data/fast/human/v_HumanRetinalIlluminance580nm_FastGroundTruthDataHistory.mat
	Fast validation      : FAILED against ground truth data of 16-Jul-2015 12:33:58.
	 > Ground truth info :             fovea.stanford.edu / MACI64, MATLAB 8.2.0.701 (R2013b) by 'wandell'
	 > Local host info   :            ray.PSYCH.UPENN.EDU / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'dhb'
	DataHash-this run    : f7f80018afa6358f66cdbb96e4c0827f24b9db9e29293ee82234a75962c22b9d
	DataHash-groundTruth : b9081882d302b5ab416aaf3959b685e6ec5bd09badb4917eedcfd7b70522dd38
	---------------------------------------------------------------------------------------------------------------------------------
	Full validation file : contains 1 instance of historical data.
	Ground truth  file   : /Users1/Shared/Dropbox/ISETBIOFullValidationData/human/v_HumanRetinalIlluminance580nm_FullGroundTruthDataHistory.mat
	Full validation      : FAILED against ground truth data of 16-Apr-2015 09:52:19.
	 > Ground truth info : ray.PSYCH.UPENN.EDU            / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'dhb'
	 > Local host info   : ray.PSYCH.UPENN.EDU            / MACI64, MATLAB 8.4.0.150421 (R2014b) by 'dhb'
	[data mismatch  1]   : 'groundTruthData.oi.optics.OTF.OTF' is a 2-D numeric whereas 'validationData.oi.optics.OTF.OTF' is a 3-D numeric.
 	[data mismatch  2]   : 'groundTruthData.oi.optics.OTF.wave' is a [ 1   1  ] matrix whereas 'validationData.oi.optics.OTF.wave' is a [31   1  ] matrix.
 	[data mismatch  3]   : Max difference between 'groundTruthData.scene.illuminant.spectrum.wave' and 'validationData.scene.illuminant.spectrum.wave' (80) is greater than the set tolerance (1.11022e-13). See figure named: ''groundTruthData.scene.illuminant.spectrum.wave' vs. 'validationData.scene.illuminant.spectrum.wave''
 	Note (*)             : script does not store any extra data.
	Validation report    : Irradiance (q/sec/m^2/nm) at 580 nm = 9.71408e+18, Expected = 1e+19, Residual = -2.85925 % is within the specified tolerance (0.03) of zero.

**Comment by npcottaris** [comment]
David Brainard said:

_Also, these scripts print a message that there are some figures that show the differences, but no such figures have appeared. I'm guessing no figures appeared because I don't have some flag set, but if so then perhaps the code that produces the message about the availability of the figures should only do so when the figures are being generated._

Actually, the difference figures are produced but the script after `v_HumanRetinalIlluminance580nm`, which is `v_OTFandPupilSize`, calls s_initISET, which eventually issues a `close all` command, thereby closing the differential figures.

To see these figures, you need to run `validateFullOne.m`

**Comment by npcottaris** [comment]
David,

1) The numerical tolerance set in the validation scripts is 500*eps, which on my machine evaluates to 
1.1102e-13. We can certainly increase this number.

2) Have you done a git pull on isetbio master branch? Last night a merged Brian's validate branch with the master branch.

**Comment by wandell** [comment]
Back to the errors that David describes - 

I see these as well.  But I considered them insignificant because the numerical values of interest were within tolerance.  I figured that the OTF data and the illuminant data are getting built differently between the original ground truth and now.  I was hoping that the rebuild of the ground truth data set would just make the size of the stored OTF and illuminant right, and we would move on.  I was just so relieved to see the numerical values of interest align that I wasn't going to fuss over whether the OTF was represented at just the one wavelength (580) or the larger array of wavelengths (400,700).  I needed to declare some kind of victory.  

By the way - s_initISET is now just a call to ieInit.  So whenever you see s_initISET replace it with ieInit.

**Comment by DavidBrainard** [comment]
v_HumanRetinalIlluminance580nm is pretty simple and still seems to validate what it was supposed to, so I went ahead and regenerated and pushed the new ground truth data.  

I think the other one is real numerical difference.

Yes, I was running with a fresh pull, and double checked just now after another one that I still get the error against Nicolas’s ground truth data.

> On Jul 30, 2015, at 4:15 PM, Brian Wandell <notifications@github.com> wrote:
> 
> Back to the errors that David describes -
> 
> I see these as well. But I considered them insignificant because the numerical values of interest were within tolerance. I figured that the OTF data and the illuminant data are getting built differently between the original ground truth and now. I was hoping that the rebuild of the ground truth data set would just make the size of the stored OTF and illuminant right, and we would move on. I was just so relieved to see the numerical values of interest align that I wasn't going to fuss over whether the OTF was represented at just the one wavelength (580) or the larger array of wavelengths (400,700). I needed to declare some kind of victory.
> 
> By the way - s_initISET is now just a call to ieInit. So whenever you see s_initISET replace it with ieInit.
> 
> —
> Reply to this email directly or view it on GitHub.
> 

David H. Brainard
Professor of Psychology
Director, Penn Vision Research Center
Director, Institute for Research in Cognitive Science
University of Pennsylvania
3401 Walnut Street, Suite 400A (mail address)
3401 Walnut Street, Room 315C (office location)
Philadelphia, PA, 19104
Phone: 215-573-7579
Email: brainard@psych.upenn.edu
Web: http://color.psych.upenn.edu/brainard

**Comment by npcottaris** [comment]
I concur that v_HumanRetinalIlluminance580nm 
passes FULL and FAST validation on my computer against David’s latest validation data.

Just to be sure, which script fails against ground truth data generated by me?

N.

**Comment by DavidBrainard** [comment]
v_DisplayColorConversion

**Comment by npcottaris** [comment]
I see the same numerical discrepancy in ptbLMSIsomerizations when I generate ground truth data on my office iMac and test against my laptop.

I thought this might be related to some difference in the PTB-3. But the validation failure persists even after upgrading PTB on both machines.

To pass the FULL validation for ptbLMSIsomerizations I had to raise the Tolerance from 500*eps to 8200*eps on my laptop.
Although the new tolerance is still a small number, 1.8208e-12, this issue remains a bit mysterious as we never had to have such a high tolerance before. I wonder is some recent PTB update computes LMSisomerizations using more machine-dependent computations.

N.

**Comment by DavidBrainard** [comment]
I don't think that code in PTB has changed for a long time.  I agree, a little mysterious that a machine dependency would suddenly appear when we haven't seen one of that size before.  But 8200*eps is still very small, and perhaps we never tested this combination of machines before.

I verified that the full validations run fine on my machine, both in publish and non publish mode.

Closing this issue.
